### ISSUE TITLE
Using custom terraform provider

### ISSUE THREAD

**Issue by bondezbond** [manheim/terraform-pipeline#205]
I have a custom terraform provider and need to download it in `~/.terraform.d/plugins` directory before running the terraform commands.

**Comment by bondezbond** [comment]
I'm thinking we make a `CustomPluginsPlugin` that reads maybe a script that downloads the custom plugins and then passes it to the commands like `-plugin-dir`

**Comment by kmanning** [comment]
Just to clarify: 

- Plugins should automatically be installed with `terraform init`.  So the only reason you would need this feature, is for a plugin that's not available on the terraform registry?  https://registry.terraform.io/
- I feel like there might be a more generic UseCase here.  Yes, you want to install a plugin manually, but more generically, you  just want to accomplish a task before you run your commands in a TerraformEnvironmentStage...?  Ideally, I think I'd like to offer a more generic solution...? (thinking).

To spitball ideas (I'm not saying this is a good thing), I think you might be able to accomplish what you want already.  It's hacky, but it helps flesh out what I'm saying above about having a  more generic UseCase.

```
@Library(['terraform-pipeline@v5.1']) _
TerraformInitComand.withPrefix('./downloadPlugin.sh; ')
Jenkinsfile.init(this)
def validate = new TerraformValidateStage()
def deployQa = new TerraformEnvironmentStage('qa')
def deployUat = new TerraformEnvironmentStage('uat')
def deployProd = new TerraformEnvironmentStage('prod')
validate.then(deployQa)
        .then(deployUat)
        .then(deployProd)
        .build()
```

If I were to spell out a less hacky  solution than what I'm showing above, I think I'd want it to look similar.  Maybe something like: 

```
@Library(['terraform-pipeline@v5.1']) _
TerraformEnvironment.before('./doMyMagic.sh')
Jenkinsfile.init(this)
def validate = new TerraformValidateStage()
def deployQa = new TerraformEnvironmentStage('qa')
def deployUat = new TerraformEnvironmentStage('uat')
def deployProd = new TerraformEnvironmentStage('prod')
validate.then(deployQa)
        .then(deployUat)
        .then(deployProd)
        .build()
```

Thoughts?

**Comment by bondezbond** [comment]
1. correct on point one. This is only for private providers that are not on the terraform registry
2. I will have to try out the hacky solution and get back to you on it. I might be able to do it with `&&` instead of `;` where it might not break the other plugins

3. The less hack solution would accomplish what we need. It can even work with the AgentNode plugin cause the script can download the plugins to a local working directory and then mount that directory to the container

```
@Library(['terraform-pipeline@v5.1']) _

TerraformEnvironment.doThisBefore('./myCustomMagic.sh')
Jenkinsfile.init(this)
AgentNodePlugin.withAgentDockerImage('hashicorp/terraform:0.10.2')
               .withAgentDockerImageOptions("--entrypoint='' -v ./terraform.d:/root/.terraform.d/")
               .init()
```

**Comment by dkolb** [comment]
🤣 Literally just came here to open an issue for @kmanning's suggestion.  And this comment turned into a wall of text.  Sorry.

I'd like to second the support of pre and post hooks in stages to allow preparation and cleanup of whatever `node` or `docker.image(image).inside(options) { }` our job ends up using.

Internally obviously we have a very large use case around a custom provider.  But there are several use cases where nodes are not completely prepared and perhaps shouldn't be.

- If using the [AgentNodePlugin](https://github.com/manheim/terraform-pipeline/blob/master/docs/AgentNodePlugin.md), any docker image will likely need to be heavily customized to include any additional dependencies.  This precludes users from just yanking the correct terraform containers directly off dockerhub, incurring the significant overhead (and probably duplication across teams) of building and storing an image somewhere and rebuilding every time the terraform version changes.

- If reusing a Jenkins slave deployment object, such as a docker image or AMI to create a long-lived slave, you end up with the same issue.  Upstream changes and the recreation of slaves can possibly remove your dependencies or recreate the slaves forcing you to manually reinstall them.  Whereas a pre-stage hook would allow you to guarantee those dependencies are always present.

That said, the interface should probably more closely resemble the decorations interface, though instead you just create a series of closures that provide Jenkins DSL acceptable to run inside `stage("pre-work") { }`.  Individual plugins can then provide convenience methods manipulate stages, such as `ShellCommandBeforeStagePlugin` specifically providing ways to prepare a node with shell commands.  (I'm bad at naming things.)

**Comment by bondezbond** [comment]
So this is mostly solved with [TerraformEnvironmentStageShellHookPlugin](./docs/TerraformEnvironmentStageShellHookPlugin.md).

One part that needs to supported is ValidationStage hook as well.

@kmanning @dkolb thoughts on closing this issue?

**Comment by kmanning** [comment]
👍  - good with me.  We can make a new issue to create a separate ValidateStageShellHookPlugin.

**Comment by bondezbond** [comment]
Released in [v5.13](https://github.com/manheim/terraform-pipeline/releases/tag/v5.13)
